sympy/core/traversal.py

```
@@ -5,6 +5,43 @@
 
 
 
+def apply(f, e, atoms=True, nonbasic=True):
+    """apply f recursively to e; if e is an iterable like list or tuple, apply
+    f recursively to the elements and return the same type of iterable.
+
+    Examples
+    ========
+
+    >>> from sympy.core.traversal import apply
+    >>> from sympy import Float, Rational, nsimplify
+    >>> from sympy.core.numbers import equal_valued
+
+    >>> apply(nsimplify, .1)
+    1/10
+    >>> apply(nsimplify, [.1])
+    [1/10]
+    >>> apply(nsimplify, [.1, [.1]])
+    [1/10, [1/10]]
+    >>> apply(nsimplify, .1, nonbasic=False)
+    0.1
+    >>> apply(nsimplify, [Float(.1)], nonbasic=False)
+    [1/10]
+    >>> recast = lambda x: y if equal_valued(x, y:=Rational(str(x))) else x
+    >>> apply(recast, [0.1,0.5,1,1.0])
+    [0.1, 1/2, 1, 1]
+    """
+    if isinstance(e, (list, tuple)):
+        return type(e)([apply(f, i, atoms, nonbasic) for i in e])
+    elif isinstance(e, Basic):
+        return bottom_up(e, f, atoms, nonbasic)
+    if nonbasic:
+        try:
+            return f(e)
+        except TypeError:
+            pass
+    return e
```

Definitely don't catch and ignore TypeError like this. Let the error bubble up.